transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java

```diff
@@ -286,23 +329,21 @@ protected void run() {
                     default:
                         // fallthrough
                 }
-
+                lastBlockingAmount.lazySet(System.nanoTime() - runStartTime);
                 final int ioRatio = this.ioRatio;
+                final long ioStartTime = System.nanoTime();
+                lastProcessedChannels.lazySet(tryProcessReady(events, strategy));
+                final long ioEndTime = System.nanoTime();
+                final long ioTime = ioEndTime - ioStartTime;
                 if (ioRatio == 100) {
-                    if (strategy > 0) {
-                        processReady(events, strategy);
-                    }
                     runAllTasks();
                 } else {
-                    final long ioStartTime = System.nanoTime();
-
-                    if (strategy > 0) {
-                        processReady(events, strategy);
-                    }
-
-                    final long ioTime = System.nanoTime() - ioStartTime;
                     runAllTasks(ioTime * (100 - ioRatio) / ioRatio);
                 }
+                final long taskTime = System.nanoTime() - ioTime;
+                lastIoProcessingTime.lazySet(ioTime * 1000);
```

Comment:
I generally favor having the metrics interface responsible for tracking the metrics. For example:

```java
public interface MetricsTracker {
   void ioProcessingBegin();
   void ioProcessingEnd();
}
```

The reasons why I generally prefer this are:

1. Provides the ability to "pull" metrics instead of "poll" for metrics changes.
  - With the existing model it forces the users to "pull"/"poll" metrics (e.g. call `lastIoProcessingTime()` at some time). However with the new model the user can use a "pull" model (e.g. by wrapping a default `MetricsTracker`, or implementing their own `MetricsTracker`), and also expose a "pull" model if this is desired.
2. Allows the user to control how metrics are tracked.
  - The user may have their own metrics system they want to persist this data to. With the current interface there is an intermediate step the user must go through to "poll" the API and translate these values to their own metric system's API. With the suggested approach a user could provide an implementation of `MetricsTracker` specifically geared for their Metrics system and thus skip this intermediate representation (and complexity of "polling").
  - The user can control exactly which metrics are tracked and doesn't have to waste resources tracking metrics that aren't meaningful to them.

@normanmaurer and I discussed this a bit but I would also like to get input from others. @nmittler @ejona86 @trustin  ... thoughts?